### PR TITLE
fix(auctions): Closed auctions fixes

### PR DESCRIPTION
**Pull request by pvinis**
The type of this PR is: **Bugfix**

This PR resolves [CX-725], [CX-728]

### Description

This includes two fixes:
- when a closed auction is accessed, it doesn't try to go live anymore.
- when a closed auction is accessed, the register button is not displayed.

### PR Checklist (tick all before merging)

- [x] I have included screenshots or videos to illustrate my changes, or I have not changed anything that impacts the UI.
- [x] I have added tests for my changes, or my changes don't require testing, or I have included a link to a separate Jira ticket covering the tests.
- [x] I have documented any follow-up work that this PR will require, or it does not require any.
- [x] I have added an app state migration, or my changes do not require one. ([What are migrations?](https://github.com/artsy/eigen/blob/master/docs/adding_state_migrations.md))
- [x] I have added a [CHANGELOG.yml](/CHANGELOG.yml) entry or my changes do not require one.

[CX-434]: https://artsyproduct.atlassian.net/browse/CX-434
[CX-725]: https://artsyproduct.atlassian.net/browse/CX-725

[CX-728]: https://artsyproduct.atlassian.net/browse/CX-728

### CODE REVIEW

File: src/lib/Scenes/Sale/Sale.tsx [hunk 1]
```diff
@@ -48,29 +51,30 @@ export const Sale: React.FC<Props> = ({ queryRes }) => {
...
+  const prevIsLive = usePrevious(isLive, false)
 
   const scrollAnim = useRef(new Animated.Value(0)).current
 
-  let intervalId: NodeJS.Timeout
+  // poll every .5 seconds to check if sale has gone live
+  useInterval(() => {
+    const now = moment()
+    setIsLive(
+      sale.liveStartAt !== null && now.isAfter(sale.liveStartAt) && sale.endAt !== null && now.isBefore(sale.endAt)
```
Comment (MounirDhahri) [inline src/lib/Scenes/Sale/Sale.tsx:63]: I think we should also make this check if the sale `endAt` is null and we're after the `liveStartAt`

File: src/lib/utils/usePrevious.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import { useEffect, useRef } from "react"
+
+export const usePrevious = <T>(value: T, initialValue: T) => {
+  const ref = useRef<T>(initialValue)
+
+  useEffect(() => {
+    ref.current = value
+  }, [value])
```
Comment (MounirDhahri) [inline src/lib/utils/usePrevious.ts:8]: Since we have value in the dependencies array I guess it's no longer the previous value right? it will be updated once the isLive value is updated and the screen is mounted

Comment (MounirDhahri) [reply src/lib/utils/usePrevious.ts:8]: **thought:** I guess you are trying to get the new `isLive` value before the screen is mounted, if that's the case I think you should move the reference and the the `useEffect`to the sale screen

**Comment by MounirDhahri** [comment]
Thanks for fixing these fast 🙏, I left few minor **non-blocking** comments. But I think we should also add a test for `when a closed auction is accessed, it doesn't try to go live anymore.` to avoid this regression in the future

**Comment by MounirDhahri** [comment]
Thanks for fixing these fast 🙏 good job. I left few minor non-blocking comments. But I think we should also add a test for _when a closed auction is accessed, it doesn't try to go live anymore._ to avoid this regression in the future

File: src/lib/Scenes/Sale/Sale.tsx (hunk 1 above)
Comment (pvinis) [reply src/lib/Scenes/Sale/Sale.tsx:63]: 🤔 makes sense.

**Comment by pvinis** [comment]
Yup, I'll add that too.

File: src/lib/Scenes/Sale/__tests__/Sale-tests.tsx [hunk 3]
```diff
@@ -78,22 +71,79 @@ describe("Sale", () => {
...
+          slug: "closed-sale-slug",
+          startAt: moment().subtract(3, "days").toISOString(),
+          liveStartAt: moment().subtract(2, "days").toISOString(),
+          endAt: moment().subtract(1, "day").toISOString(),
+          name: "closed sale!",
+        }),
+      })
+    )
+
+    expect(tree.findAllByType(RegisterToBidButton)).toHaveLength(0)
```
Comment (brainbicycle) [inline src/lib/Scenes/Sale/__tests__/Sale-tests.tsx:147]: 🎉

File: src/lib/Scenes/Sale/Sale.tsx [hunk 4]
```diff
@@ -48,29 +51,30 @@ export const Sale: React.FC<Props> = ({ queryRes }) => {
...
+    )
+  }, 500)
```
Comment (brainbicycle) [inline src/lib/Scenes/Sale/Sale.tsx:65]: nice new useInterval is cleaner ✨

**Review by brainbicycle — APPROVED** [review]

File: src/lib/utils/usePrevious.ts (hunk 2 above)
Comment (pvinis) [reply src/lib/utils/usePrevious.ts:8]: Mm I'm not sure what you mean. I used https://usehooks.com/usePrevious/ and modified it to have an initial value. The return is happening before the useEffect, so the usePrevious hook always gives us the previous value. What I use it for is to know when `isLive` is `true` and `prevIsLive` is `false`. That's the only case where we should navigate to the live screen. In any other case we do nothing.

**Comment by pvinis** [comment]
This should be ready. I'll add @admbtlr as a reviewer because I changed stuff in the `saleTime`. I tried to make types strict, and ended up with a couple of edge cases that were missing. I talked with Sam for how to handle these, changed the tests to reflect that, and then changed the code to make them all pass.

File: src/lib/Scenes/Sale/Sale.tsx [hunk 5]
```diff
@@ -8,6 +8,9 @@ import { defaultEnvironment } from "lib/relay/createEnvironment"
 import { getCurrentEmissionState } from "lib/store/AppStore"
 import { ArtworkFilterContext, ArtworkFilterGlobalStateProvider } from "lib/utils/ArtworkFilter/ArtworkFiltersStore"
 import { Schema } from "lib/utils/track"
+import { useInterval } from "lib/utils/useInterval"
+import { usePrevious } from "lib/utils/usePrevious"
+import _ from "lodash"
```
Comment (admbtlr) [inline src/lib/Scenes/Sale/Sale.tsx:13]: Shouldn't this rather be 
```
import compact from 'lodash/compact'
```
in order to avoid importing the whole caboodle?

File: src/lib/Scenes/Sale/Sale.tsx [hunk 6]
```diff
@@ -48,29 +51,45 @@ export const Sale: React.FC<Props> = ({ queryRes }) => {
...
-    if (sale.liveStartAt) {
-      // poll every .5 seconds to check if sale has gone live
-      intervalId = setInterval(checkIfSaleIsLive, 500)
-      return () => {
-        clearInterval(intervalId)
-      }
+  // poll every .5 seconds to check if sale has gone live
+  useInterval(() => {
+    const now = moment()
+    if (sale.liveStartAt === null) {
```
Comment (admbtlr) [inline src/lib/Scenes/Sale/Sale.tsx:62]: If it's null, then we don't need to do the `useInterval` in the first place, do we? Unless we're expecting that value to change from null at some point?

File: src/lib/Scenes/Sale/helpers/__tests__/auctionTime-tests.ts [hunk 7]
```diff
@@ -1,105 +1,227 @@
-// import moment from "moment-timezone"
-
-const TIME_ZONE_NY = "America/New_York"
-const TIME_ZONE_LA = "America/Los_Angeles"
-const TIME_ZONE_DE = "Europe/Berlin"
+const timezones = {
```
Comment (admbtlr) [inline src/lib/Scenes/Sale/helpers/__tests__/auctionTime-tests.ts:1]: Nice work on these tests!

File: src/lib/Scenes/Sale/helpers/saleTime.ts [hunk 8]
```diff
@@ -1,68 +1,122 @@
...
+  if (startDateMoment !== null && now.isBefore(startDateMoment)) {
+    return begins(startDateMoment, userTimeZone, saleType)
+  }
+
+  // definitely already closed
+  if (endDateMoment !== null && now.isAfter(endDateMoment)) {
+    return closed(endDateMoment)
+  }
+
+  // if we have both start and end and we in the middle
```
Comment (admbtlr) [inline src/lib/Scenes/Sale/helpers/saleTime.ts:45]:
```suggestion
  // if we have both start and end and we're in between them
```

File: src/lib/utils/usePrevious.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import { useEffect, useRef } from "react"
```
Comment (admbtlr) [inline src/lib/utils/usePrevious.ts:1]: Do we need this util? This is a known pattern, and hiding it in a 6 line library function is kinda obfuscating

**Review by admbtlr — CHANGES_REQUESTED** [review]

File: src/lib/Scenes/Sale/Sale.tsx (hunk 5 above)
Comment (pvinis) [reply src/lib/Scenes/Sale/Sale.tsx:13]: Nothing really changes in RN. We are not web (thankfully, in this case 😅). In most files we import like that, and since there is no change in perf or whatever, why not keep it consistent and simple. 👍

File: src/lib/Scenes/Sale/helpers/saleTime.ts (hunk 8 above)
Comment (pvinis) [reply src/lib/Scenes/Sale/helpers/saleTime.ts:45]: 🇬🇧

File: src/lib/utils/usePrevious.ts (hunk 9 above)
Comment (pvinis) [reply src/lib/utils/usePrevious.ts:1]: 🤔 I'm not sure I understand 😬. What do you mean?

The `usePrevious` is kind of a nice and helpful hook. Apparently we are not using it anywhere in eigen already, so I thought I'll add this here.

File: src/lib/Scenes/Sale/Sale.tsx (hunk 6 above)
Comment (pvinis) [reply src/lib/Scenes/Sale/Sale.tsx:62]: Indeed. That was one of my thoughts against using useInterval. But then again when I saw the useEffect with an if and an interval and cleanup return func inside that if, I felt confused and a bit uncertain of how that works, and worse, if/how that could break.
So I decided to flip that and run the interval outside the if, and have the if check inside.

This means the interval now will run even if it's not a live auction and the code is simpler. I felt I'm ok with that tradeoff.

Comment (pvinis) [inline src/lib/Scenes/Sale/Sale.tsx:62]:
```suggestion
    if (sale.liveStartAt === null) {
      const now = moment()
```
I'll quickly do this. This feels slightly nicer.

File: src/lib/Scenes/Sale/Sale.tsx [hunk 10]
```diff
@@ -48,29 +51,45 @@ export const Sale: React.FC<Props> = ({ queryRes }) => {
...
+      setIsLive(false)
+      return
     }
```
Comment (pvinis) [inline src/lib/Scenes/Sale/Sale.tsx:65]:
```suggestion
    if (sale.liveStartAt === null) {
      setIsLive(false)
      return
    }
    const now = moment()
```
I'll do this. This feels slightly nicer.

File: src/lib/utils/usePrevious.ts (hunk 2 above)
Comment (MounirDhahri) [reply src/lib/utils/usePrevious.ts:8]: Sorry, I must have misunderstood the effect then 😬

File: src/lib/utils/usePrevious.ts (hunk 9 above)
Comment (admbtlr) [reply src/lib/utils/usePrevious.ts:1]: Actually, you're right, creating a named hook to do this is a nice approach 👍 . (I was just thinking that it was unnecessary to create a separate hook in order to abstract away 6 lines of code)

Comment (pvinis) [reply src/lib/utils/usePrevious.ts:1]: Aaah gotcha. Thanks for explaining. Yes, I think it's nice to have, and convenient to have it as a reusable hook.

**Comment by MounirDhahri** [comment]
Thanks for adding the tests and the quick fixes 🌷

**Review by MounirDhahri — APPROVED** [review]

**Review by admbtlr — APPROVED** [review]
